feat(data_classes): Add CloudWatchAlarmEvent data class

par6n opened the pull request:
<!-- markdownlint-disable MD041 MD043 -->
**Issue number: #3848**

## Summary
AWS has recently introduced AWS Lambda as an alarm state change action ([link to the news entry](https://aws.amazon.com/about-aws/whats-new/2023/12/amazon-cloudwatch-alarms-lambda-change-action/)). This pull request facilitates reading the event that a CloudWatch Alarm sends to a Lambda function.

### Changes

> Please provide a summary of what's being changed
* Introduces a new data class: `CloudWatchAlarmEvent`, along with its satellite data classes and enums.

Pending items:
* [x] Test the class against complex alarms, for example, the ones with mathematical expressions.

### User experience

> Please share what the user experience looks like before and after this change

You will be able to deserialize the CloudWatch Alarm state change event. Docstrings are added for fields.

```python
from aws_lambda_powertools.utilities.data_classes import event_source, CloudWatchAlarmEvent

@event_source(data_class=CloudWatchAlarmEvent)
def lambda_handler(event: CloudWatchAlarmEvent, context):
    if event.state.value.name == 'ALARM':
        print(f"{event.alarm_arn} is on alarm...")
```

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

* [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
* [x] I have performed a self-review of this change
* [x] Changes have been tested
* [x] Changes are documented
* [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

* [ ] Migration process documented
* [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

par6n committed:
feat(data_classes): initialise cloud_watch_alarm_event data class

par6n committed:
fix(data_classes): add CloudWatchAlarmEvent to dataclasses index

par6n committed:
fix(data_classes): add the satellite classes to index

rubenfonseca committed:
Merge branch 'develop' into feat/add-cloud-watch-alarm-event-data-class

par6n committed:
fix(data_classes): address mypy issues

par6n committed:
docs(data_classes): add documentation on CloudWatchAlarmEvent

par6n committed:
fix(data_classes): add 'expression' and 'label' fields to CloudWatchAlarmMetric

par6n commented:
@rubenfonseca I'm done with the changes. I've tried running `make lint-docs`, however it exits with code zero, without returning any errors. I suppose that means no linting issues were found?

leandrodamascena commented:
Looking at this

leandrodamascena commented on aws_lambda_powertools/utilities/data_classes/cloud_watch_alarm_event.py:
```diff
@@ -0,0 +1,206 @@
+
+class CloudWatchAlarmState(DictWrapper):
+    @property
+    def value(self) -> CloudWatchAlarmStateValue:
```
Instead of returning an Enum, it's better to return a Literal directly. The values are always `OK`, `ALARM` or `INSUFFICIENT_DATA` and a Literal makes this more explicit without having to create a new Enum class for it.

par6n replied:
Ah, coming from the TypeScript world, Python literals make me happy 😄 .

leandrodamascena commented on aws_lambda_powertools/utilities/data_classes/cloud_watch_alarm_event.py:
```diff
@@ -0,0 +1,206 @@
+        return self.get("reason")
+
+    @property
+    def reason_data(self) -> Optional[dict]:
+        """
+        Additional data to back up the reason, usually contains the evaluated data points,
+        the calculated threshold and timestamps.
+        """
+        if self.get("reasonData") is None:
+            return None
+
+        return json.loads(str(self.get("reasonData")))
```
We have 2 things to do here:

1 - It is good to have a field that returns the raw value, that is, customers may not want to deserialize the value and want to use the raw string. So keep this field just returning `self.get("reasonData")` and create a new one called `reason_data_decoded`
2 - In this new field, allow the customer to bring their own json deserializer, as we do here https://github.com/aws-powertools/powertools-lambda-python/blob/develop/aws_lambda_powertools/utilities/data_classes/active_mq_event.py#L28

leandrodamascena commented on aws_lambda_powertools/utilities/data_classes/cloud_watch_alarm_event.py:
```diff
@@ -0,0 +1,206 @@
+        return self["time"]
+
+    @property
+    def alarm_name(self) -> str:
+        """
+        Alarm name.
+        """
+        return self["alarmData"]["alarmName"]
```
I think we need to create a new `AlarmData` class and then customers will be able to access `event.alarm_data.alarm_name`. It is valid for all fields nested in the `alarm_data` key. E.g `event.alarm_data.state.value`.

leandrodamascena commented on docs/utilities/data_classes.md:
````diff
@@ -528,6 +529,23 @@ In this example, we also use the new Logger `correlation_id` and built-in `corre
+[CloudWatch supports Lambda as an alarm state change action](https://aws.amazon.com/about-aws/whats-new/2023/12/amazon-cloudwatch-alarms-lambda-change-action/){target="_blank"}.
+You can use the `CloudWathAlarmEvent` data class to access the fields containing such data as alarm information, current state, and previous state.
+
+=== "app.py"
+
+    ```python
+    from aws_lambda_powertools.utilities.data_classes import event_source, CloudWatchAlarmEvent
+
+    @event_source(data_class=CloudWatchAlarmEvent)
+    def lambda_handler(event: CloudWatchAlarmEvent, context):
+        if event.state.value.name == "ALARM":
+            print(f"{event.alarm_name} is on alarm because {event.state.reason}...")
+            do_something_with(event.alarm_arn)
+    ```
+
````
Please move this example to `examples/event_sources`.

par6n replied:
Done, however, I wasn't sure whether to remove the example from the documentation or not.

leandrodamascena replied:
Yes, you can remove the code from here and include the file like this https://github.com/aws-powertools/powertools-lambda-python/blob/develop/docs/utilities/data_classes.md?plain=1#L999

par6n replied:
Done, thanks!

leandrodamascena commented on tests/events/cloudWatchAlarmEvent.json:
```diff
@@ -0,0 +1,59 @@
+{
+  "source": "aws.cloudwatch",
+  "alarmArn": "arn:aws:cloudwatch:eu-west-1:912397435824:alarm:test_alarm",
+  "accountId": "000000000000",
```
```suggestion
  "accountId": "123456789012",
```

leandrodamascena reviewed (changes requested):
Hello @par6n! Thank you for sending this OR! This is great work.

I made some comments to improve the customer experience and hope we can address them soon and merge this PR!

par6n committed:
fix(data_classes): change accountId to `123456789012` in cloudWatchAlarmEvent sample

par6n committed:
fix(data_classes): add a new `reason_data_decoded` property to CloudWatchAlarmState

This will allow the customers to easily access the raw 'reasonData' field, as well as accessing parsed data using the new property.

par6n committed:
fix(data_classes): use Literal instead of Enum for the property `value` in CloudWatchAlarmState

par6n committed:
improv(data_classes): introduce CloudWatchAlarmData data class that contains the alarm data

par6n committed:
improv(data_classes): change source property return type to Literal

par6n committed:
docs(data_classes): update the example for CloudWatch Alarm State Change Action

par6n committed:
docs(data_classes): add a working example under `examples/event_sources/`

par6n commented:
@leandrodamascena Thanks for the excellent comments! I'm not a Python-savvy, and I've already learned a few things working on this.

par6n committed:
improv(data_classes): use `cached_property` decorator for `reason_data_decoded`

par6n committed:
docs(data_classes): reformat table in data_classes

leandrodamascena commented:
> @leandrodamascena Thanks for the excellent comments! I'm not a Python-savvy, and I've already learned a few things working on this.

Ohhh really? No one can say that you haven't written code in Python for a long time, the code is very clean and well documented.

I'm checking the documentation and I see some fields that are missing, for example when you have a composite alarm you have additional fields, for example:

```json
'previousState': {
      'value': 'ALARM',
      'reason': 'arn:aws:cloudwatch:us-east-1:111122223333:alarm:CompositeDemo.FirstChild transitioned to ALARM at Friday 04 August, 2023 12:54:46 UTC',
      'reasonData': '{"triggeringAlarms":[{"arn":"arn:aws:cloudwatch:us-east-1:111122223333:alarm:CompositeDemo.FirstChild","state":{"value":"ALARM","timestamp":"2023-08-04T12:54:46.138+0000"}}]}',
      'timestamp': '2023-08-04T12:54:46.138+0000',
      'actionsSuppressedBy': 'WaitPeriod',
      'actionsSuppressedReason': 'Actions suppressed by WaitPeriod'
    },
```

It's very late here and I'm wrapping up the day. Please give me until tomorrow to push some commits that address certain issues. I will explain the details of the fixes that I address.

Please let me know if you have any other questions in the meantime.

Thanks again.

leandrodamascena committed:
Merge branch 'develop' into feat/add-cloud-watch-alarm-event-data-class

par6n committed:
docs(data_classes): fix cloudwatch_alarm_event example typing issue

par6n committed:
docs(data_classes): replace example code with reference to the example file for CW alarm state change action

par6n committed:
Merge remote-tracking branch 'origin/feat/add-cloud-watch-alarm-event-data-class' into feat/add-cloud-watch-alarm-event-data-class

par6n committed:
feat(data_classes): add `actions_suppressed_by` and `actions_suppressed_reason` properties to CloudWatchAlarmState

rubenfonseca committed:
Merge branch 'develop' into feat/add-cloud-watch-alarm-event-data-class

leandrodamascena committed:
Merge branch 'develop' into feat/add-cloud-watch-alarm-event-data-class

leandrodamascena committed:
Refactoring code

leandrodamascena committed:
Refactoring code

leandrodamascena commented:
Hello @par6n! I pushed a commit with a fix for a few small things and here is the list of what I changed:

1 - I removed the `CloudWatchAlarmStateValue` and `CloudWatchAlarmActionSuppressor` variables. We are only using this once and can add the type annotation directly to the property.
2 - I optimized some "ifs" to make the code cleaner.
3 - Our intention is to allow the client to access the properties following the same structure they have in JSON, therefore, I changed the location of some fields and created two new classes: `CloudWatchAlarmMetricStat` and `CloudWatchAlarmConfiguration`.
4 - I modified the test a little to make it more realistic.

Thank you for your outstanding work on this task. Before merging, I'm sending this PR to @rubenfonseca for review. If you have any questions or need clarification, please feel free to reach out to me.

par6n commented:
@leandrodamascena I checked out your changes, and it all looks good to me. Excellent work over there! If there's anything I could do further for this PR, please let me know!

rubenfonseca commented:
Reviewing now

leandrodamascena reviewed (approved):
APPROVED!! @par6n
